File: flutter/flet_spinkit/lib/src/spinkit.dart
import 'package:flet/flet.dart';
import 'package:flutter/material.dart';
import 'package:flutter_spinkit/flutter_spinkit.dart';

class SpinkitControl extends StatelessWidget {
  final Control? parent;
  final Control control;

  const SpinkitControl({
    super.key,
    required this.parent,
    required this.control,
  });

  @override
  Widget build(BuildContext context) {
    var color = control.attrColor("color", context);
    var size = control.attrDouble("size", 50)!;
    var spinkitType = control.attrString("spinkitType");
    var duration = parseDuration(control, "duration");
    late Widget spinkitWidget;

    switch (spinkitType?.toLowerCase()) {
      case "rotatingcircle":
        spinkitWidget = SpinKitRotatingCircle(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 1200),
        );
        break;
      case "foldingcube":
        spinkitWidget = SpinKitFoldingCube(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 2400),
        );
        break;
      case "doublebounce":
        spinkitWidget = SpinKitDoubleBounce(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 2000),
        );
        break;
      case "wave":
        spinkitWidget = SpinKitWave(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 1200),
        );
        break;
      case "wanderingcubes":
        spinkitWidget = SpinKitWanderingCubes(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 1800),
        );
        break;
      case "circle":
        spinkitWidget = SpinKitCircle(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 1200),
        );
        break;
      case "fadingcircle":
        spinkitWidget = SpinKitFadingCircle(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 1200),
        );
        break;
      case "dancingsquare":
        spinkitWidget = SpinKitDancingSquare(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 1200),
        );
        break;
      case "chasingdots":
        spinkitWidget = SpinKitChasingDots(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 2000),
        );
        break;
      case "threebounce":
        spinkitWidget = SpinKitThreeBounce(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 1400),
        );
        break;
      case "fadingfour":
        spinkitWidget = SpinKitFadingFour(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 1200),
        );
        break;
      case "pumpingheart":
      default:
        spinkitWidget = SpinKitPumpingHeart(
          color: color,
          size: size,
          duration: duration ?? const Duration(milliseconds: 2400),
        );
    }

    return constrainedControl(context, spinkitWidget, parent, control);
  }
}

File: python/flet_spinkit/flet_spinkit/spinkit.py
from enum import Enum
from typing import Any, Optional

from flet_core.constrained_control import ConstrainedControl
from flet_core.control import OptionalNumber
from flet_core.types import DurationValue


class SpinkitType(Enum):
    ROTATING_CIRCLE = "rotatingCircle"
    FOLDING_CUBE = "foldingCube"
    DOUBLE_BOUNCE = "doubleBounce"
    WAVE = "wave"
    WANDERING_CUBES = "wanderingCubes"
    CIRCLE = "circle"
    FADING_CIRCLE = "fadingCircle"
    DANCING_SQUARE = "dancingSquare"
    CHASING_DOTS = "chasingDots"
    THREE_BOUNCE = "threeBounce"
    FADING_FOUR = "fadingFour"
    PUMPING_HEART = "pumpingHeart"


class Spinkit(ConstrainedControl):
    """
    Spinkit Control.
    """

    def __init__(
        self,
        #
        # Spinkit specific
        #
        spinkit_type: SpinkitType = SpinkitType.PUMPING_HEART,
        color: Optional[str] = None,
        size: OptionalNumber = None,
        duration: DurationValue = None,
        #
        # Control
        #
        opacity: OptionalNumber = None,
        tooltip: Optional[str] = None,
        visible: Optional[bool] = None,
        data: Any = None,
        #
        # ConstrainedControl
        #
        left: OptionalNumber = None,
        top: OptionalNumber = None,
        right: OptionalNumber = None,
        bottom: OptionalNumber = None,
    ):
        ConstrainedControl.__init__(
            self,
            tooltip=tooltip,
            opacity=opacity,
            visible=visible,
            data=data,
            left=left,
            top=top,
            right=right,
            bottom=bottom,
        )

        self.color = color
        self.size = size
        self.spinkit_type = spinkit_type
        self.duration = duration

    def _get_control_name(self):
        return "spinkit"

    def before_update(self):
        super().before_update()
        self._set_attr("duration", self.__duration)

    # color
    @property
    def color(self) -> Optional[str]:
        return self._get_attr("color")

    @color.setter
    def color(self, value: Optional[str]):
        self._set_attr("color", value)

    # size
    @property
    def size(self) -> OptionalNumber:
        return self._get_attr("size", data_type="float")

    @size.setter
    def size(self, value: OptionalNumber):
        self._set_attr("size", value)

    # spinkit_type
    @property
    def spinkit_type(self) -> SpinkitType:
        return self.__spinkit_type

    @spinkit_type.setter
    def spinkit_type(self, value: SpinkitType):
        self.__spinkit_type = value
        self._set_attr(
            "spinkitType", value.value if isinstance(value, SpinkitType) else value
        )

    # duration
    @property
    def duration(self) -> DurationValue:
        return self.__duration

    @duration.setter
    def duration(self, value: DurationValue):
        self.__duration = value
